llarp/iwp/session.hpp

```diff
@@ -0,0 +1,189 @@
+#ifndef LLARP_IWP_SESSION_HPP
+#define LLARP_IWP_SESSION_HPP
+
+#include <link/session.hpp>
+#include <iwp/linklayer.hpp>
+#include <iwp/message_buffer.hpp>
+
+namespace llarp
+{
+  namespace iwp
+  {
+    struct Session : public ILinkSession,
+                     public std::enable_shared_from_this< Session >
+    {
+      /// outbound session
+      Session(LinkLayer* parent, RouterContact rc, AddressInfo ai);
+      /// inbound session
+      Session(LinkLayer* parent, Addr from);
+
+      ~Session();
+
+      void
+      Pump() override;
+
+      void
+      Tick(llarp_time_t now) override;
+
+      bool
+      SendMessageBuffer(const llarp_buffer_t& buf,
+                        CompletionHandler resultHandler) override;
+
+      void
+      Send_LL(const llarp_buffer_t& pkt);
+
+      void
+      EncryptAndSend(const llarp_buffer_t& data);
+
+      void
+      Start() override;
+
+      void
+      Close() override;
+
+      void
+      Recv_LL(const llarp_buffer_t& pkt) override;
+
+      bool
+      SendKeepAlive() override;
+
+      bool
+      IsEstablished() const override;
+
+      bool
+      TimedOut(llarp_time_t now) const override;
+
+      PubKey
+      GetPubKey() const override
+      {
+        return m_RemoteRC.pubkey;
+      }
+
+      Addr
+      GetRemoteEndpoint() const override
+      {
+        return m_RemoteAddr;
+      }
+
+      RouterContact
+      GetRemoteRC() const override
+      {
+        return m_RemoteRC;
+      }
+
+      size_t
+      SendQueueBacklog() const override
+      {
+        return m_TXMsgs.size();
+      }
+
+      ILinkLayer*
+      GetLinkLayer() const override
+      {
+        return m_Parent;
+      }
+
+      bool
+      RenegotiateSession() override;
+
+      bool
+      ShouldPing() const override;
+
+      util::StatusObject
+      ExtractStatus() const override;
+
+     private:
+      enum class State
```

Comment:
the design here could probably be reworked to look a bit more like https://khuttun.github.io/2017/02/04/implementing-state-machines-with-std-variant.html, but this matches the current design of the stuff.